### PR TITLE
Adding a refresh button to clear cache and reload

### PR DESCRIPTION
**Pull request by MalayaliRobz**
Introducing a reload cache button that addresses https://github.com/coronasafe/care_fe/issues/393

image

**working:**
added a rebuild script that generates 'build-meta.json' to the public folder. 'build-meta' contains a version no. specific to that build.
Now when the user clicks on the update button if the version no stored in his/her browser localstorage is not equal to the one from 'build-meta' we clear cache and reload browser.

But you may be asking but why store version in build-meta file, why not just clear cache and reload all the time.
Glad you asked.

Let's say we do a frontend deployment. But that change will reflect for the user only if he/she does a refresh of the page. But there may be scenarios where there are users that don't refresh the page for a very long time. For them, things might stop working and they will have no clue.

The idea is to show some kinda a popup or notification that there is a new version available for these users. That says "hay there is a new version pls reload". Something like what you get in facebook pages while browsing.
Didn't implement this. This PR is a base of that.

### CODE REVIEW

**Comment by tomahawk-pilot** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 19
           

Complexity increasing per file
==============================
- build-scripts/generate-build-version.js  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957&bid=19114748)

File: build-scripts/generate-build-version.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+fs.writeFile(
+  './public/build-meta.json',
+  jsonContent,
+  { flag: 'w+', encoding: 'utf8' },
+  err => {
+    if (err) {
+      console.log('An error occured while writing JSON Object to meta.json');
+      return console.log(err);
+    }
+    console.log('build-meta.json file has been saved with latest version number');
```
Comment (tomahawk-pilot) [inline build-scripts/generate-build-version.js:23]: Codacy found an issue: [Unexpected console statement.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: build-scripts/generate-build-version.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+const jsonContent = JSON.stringify(jsonData);
+
+fs.writeFile(
+  './public/build-meta.json',
+  jsonContent,
+  { flag: 'w+', encoding: 'utf8' },
+  err => {
+    if (err) {
+      console.log('An error occured while writing JSON Object to meta.json');
+      return console.log(err);
```
Comment (tomahawk-pilot) [inline build-scripts/generate-build-version.js:21]: Codacy found an issue: [Unexpected console statement.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: src/Components/Users/UserProfile.tsx [hunk 3]
```diff
@@ -186,9 +187,33 @@ export default function UserProfile() {
...
     };
 
-    return (
+    const checkForNewBuildVersion = async () => {
+
+        let [isLatestBundle, newVersion] = await checkIfLatestBundle();
+
+        if (!isLatestBundle) {
+            setUpdateBtnText("updating...");
+            localStorage.setItem('build_meta_version', newVersion);
```
Comment (tomahawk-pilot) [inline src/Components/Users/UserProfile.tsx:196]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: src/Components/Users/UserProfile.tsx [hunk 4]
```diff
@@ -186,9 +187,33 @@ export default function UserProfile() {
...
+
+            if ('caches' in window) {
```
Comment (tomahawk-pilot) [inline src/Components/Users/UserProfile.tsx:198]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: src/Components/Users/UserProfile.tsx [hunk 5]
```diff
@@ -186,9 +187,33 @@ export default function UserProfile() {
...
+                // Service worker cache should be cleared with caches.delete()
+                caches.keys().then(names => {
```
Comment (tomahawk-pilot) [inline src/Components/Users/UserProfile.tsx:200]: Codacy found an issue: [Expected parentheses around arrow function argument.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: build-scripts/generate-build-version.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+const jsonData = {
+  version: appVersion
+};
+
+const jsonContent = JSON.stringify(jsonData);
+
+fs.writeFile(
+  './public/build-meta.json',
+  jsonContent,
+  { flag: 'w+', encoding: 'utf8' },
```
Comment (tomahawk-pilot) [inline build-scripts/generate-build-version.js:17]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: build-scripts/generate-build-version.js (hunk 1 above)
Comment (tomahawk-pilot) [inline build-scripts/generate-build-version.js:23]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: src/Components/Users/UserProfile.tsx [hunk 7]
```diff
@@ -9,7 +9,7 @@ import { PhoneNumberField, SelectField, TextInputField } from "../Common/HelperI
 import { parsePhoneNumberFromString } from 'libphonenumber-js';
 import { validateEmailAddress, phonePreg } from "../../Common/validation";
 import * as Notification from "../../Utils/Notifications.js";
-
+import { checkIfLatestBundle } from '../../Utils/build-meta-info'
```
Comment (tomahawk-pilot) [inline src/Components/Users/UserProfile.tsx:12]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: build-scripts/generate-build-version.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  err => {
+    if (err) {
+      console.log('An error occured while writing JSON Object to meta.json');
```
Comment (tomahawk-pilot) [inline build-scripts/generate-build-version.js:20]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: build-scripts/generate-build-version.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+  version: appVersion
+};
+
+const jsonContent = JSON.stringify(jsonData);
+
+fs.writeFile(
+  './public/build-meta.json',
+  jsonContent,
+  { flag: 'w+', encoding: 'utf8' },
+  err => {
```
Comment (tomahawk-pilot) [inline build-scripts/generate-build-version.js:18]: Codacy found an issue: [Expected parentheses around arrow function argument.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: src/Components/Users/UserProfile.tsx [hunk 10]
```diff
@@ -186,9 +187,33 @@ export default function UserProfile() {
...
+                    }
+
+                    window.location.reload(true);
+                });
+            }
+        } else {
+            setUpdateBtnText("You already have the latest version!")
 
-        <div >
+            setTimeout(() => setUpdateBtnText("Update"), 1000);
```
Comment (tomahawk-pilot) [inline src/Components/Users/UserProfile.tsx:211]: Codacy found an issue: [The function setTimeout can be unsafe](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: build-scripts/generate-build-version.js [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+const appVersion = uuidv4();
+
+const jsonData = {
+  version: appVersion
+};
+
+const jsonContent = JSON.stringify(jsonData);
+
+fs.writeFile(
+  './public/build-meta.json',
```
Comment (tomahawk-pilot) [inline build-scripts/generate-build-version.js:15]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: build-scripts/generate-build-version.js [hunk 12]
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env node
+
+const fs = require('fs');
+const { v4: uuidv4 } = require('uuid');
```
Comment (tomahawk-pilot) [inline build-scripts/generate-build-version.js:4]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: src/Components/Users/UserProfile.tsx (hunk 7 above)
Comment (tomahawk-pilot) [inline src/Components/Users/UserProfile.tsx:12]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: src/Components/Users/UserProfile.tsx [hunk 13]
```diff
@@ -186,9 +187,33 @@ export default function UserProfile() {
...
+        }
+    }
```
Comment (tomahawk-pilot) [inline src/Components/Users/UserProfile.tsx:213]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: src/Components/Users/UserProfile.tsx [hunk 14]
```diff
@@ -186,9 +187,33 @@ export default function UserProfile() {
...
+                caches.keys().then(names => {
+                    for (const name of names) {
+                        caches.delete(name);
+                    }
+
+                    window.location.reload(true);
+                });
+            }
+        } else {
+            setUpdateBtnText("You already have the latest version!")
```
Comment (tomahawk-pilot) [inline src/Components/Users/UserProfile.tsx:209]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: build-scripts/generate-build-version.js [hunk 15]
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env node
+
+const fs = require('fs');
```
Comment (tomahawk-pilot) [inline build-scripts/generate-build-version.js:3]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: src/Components/Users/UserProfile.tsx (hunk 3 above)
Comment (tomahawk-pilot) [inline src/Components/Users/UserProfile.tsx:196]: Codacy found an issue: [localStorage can be unsafe](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: build-scripts/generate-build-version.js (hunk 8 above)
Comment (tomahawk-pilot) [inline build-scripts/generate-build-version.js:20]: Codacy found an issue: [Unexpected console statement.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

**Comment by dauntlessnomad** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 18
           

Complexity increasing per file
==============================
- build-scripts/generate-build-version.js  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957&bid=19114748)

File: src/Components/Users/UserProfile.tsx (hunk 3 above)
Comment (dauntlessnomad) [inline src/Components/Users/UserProfile.tsx:196]: Codacy found an issue: [localStorage can be unsafe](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

Comment (dauntlessnomad) [inline src/Components/Users/UserProfile.tsx:196]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: build-scripts/generate-build-version.js (hunk 12 above)
Comment (dauntlessnomad) [inline build-scripts/generate-build-version.js:4]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: build-scripts/generate-build-version.js (hunk 15 above)
Comment (dauntlessnomad) [inline build-scripts/generate-build-version.js:3]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: build-scripts/generate-build-version.js (hunk 1 above)
Comment (dauntlessnomad) [inline build-scripts/generate-build-version.js:23]: Codacy found an issue: [Unexpected console statement.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: src/Components/Users/UserProfile.tsx (hunk 14 above)
Comment (dauntlessnomad) [inline src/Components/Users/UserProfile.tsx:209]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: build-scripts/generate-build-version.js (hunk 6 above)
Comment (dauntlessnomad) [inline build-scripts/generate-build-version.js:17]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: src/Components/Users/UserProfile.tsx (hunk 10 above)
Comment (dauntlessnomad) [inline src/Components/Users/UserProfile.tsx:211]: Codacy found an issue: [The function setTimeout can be unsafe](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: src/Components/Users/UserProfile.tsx (hunk 13 above)
Comment (dauntlessnomad) [inline src/Components/Users/UserProfile.tsx:213]: Codacy found an issue: [Missing semicolon.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: build-scripts/generate-build-version.js (hunk 8 above)
Comment (dauntlessnomad) [inline build-scripts/generate-build-version.js:20]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: build-scripts/generate-build-version.js (hunk 11 above)
Comment (dauntlessnomad) [inline build-scripts/generate-build-version.js:15]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: src/Components/Users/UserProfile.tsx (hunk 4 above)
Comment (dauntlessnomad) [inline src/Components/Users/UserProfile.tsx:198]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: build-scripts/generate-build-version.js (hunk 1 above)
Comment (dauntlessnomad) [inline build-scripts/generate-build-version.js:23]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: src/Components/Users/UserProfile.tsx [hunk 16]
```diff
@@ -9,7 +9,7 @@ import { PhoneNumberField, SelectField, TextInputField } from "../Common/HelperI
 import { parsePhoneNumberFromString } from 'libphonenumber-js';
 import { validateEmailAddress, phonePreg } from "../../Common/validation";
 import * as Notification from "../../Utils/Notifications.js";
-
+import { checkIfLatestBundle } from '../../Utils/build-meta-info';
```
Comment (dauntlessnomad) [inline src/Components/Users/UserProfile.tsx:12]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: build-scripts/generate-build-version.js (hunk 8 above)
Comment (dauntlessnomad) [inline build-scripts/generate-build-version.js:20]: Codacy found an issue: [Unexpected console statement.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: src/Components/Users/UserProfile.tsx (hunk 5 above)
Comment (dauntlessnomad) [inline src/Components/Users/UserProfile.tsx:200]: Codacy found an issue: [Expected parentheses around arrow function argument.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: build-scripts/generate-build-version.js (hunk 9 above)
Comment (dauntlessnomad) [inline build-scripts/generate-build-version.js:18]: Codacy found an issue: [Expected parentheses around arrow function argument.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

File: build-scripts/generate-build-version.js (hunk 2 above)
Comment (dauntlessnomad) [inline build-scripts/generate-build-version.js:21]: Codacy found an issue: [Unexpected console statement.](https://app.codacy.com/gh/coronasafe/care_fe/pullRequest?prid=5837957)

**Pull request merged**